Medium: [Crystallizers] MSMPR steady-state solve passes scalar concentration to vector kinetics

## Maintainer triage (June 29, 2026)
- **Status:** Verified bug against the current `master` source in this fork.
- **Severity:** Medium.
- **Area labels:** area:crystallizers, area:kinetics.
- **Tackle batch:** P5 - medium-scope correctness defects.

The source review confirmed that the cited code path and failure mode are still present. The original audit details are preserved below for reproduction notes and suggested fixes.

---

## Original audit details

## Finding
`MSMPR.solve_steady_state` passes a scalar tank mass fraction into `CrystKinetics.get_kinetics`, but `get_kinetics` assumes an array with a transpose and target index.

## Evidence
- `MSMPR.solve_steady_state` calls `self.Kinetics.get_kinetics(w_tank, temp, kv)` at `PharmaPy/Crystallizers.py:1774-1777`.
- `CrystKinetics.get_kinetics` immediately evaluates `conc.T[self.target_idx]` at `PharmaPy/Kinetics.py:826-833`.
- Scalars do not support this indexing; numpy scalar inputs raise `IndexError` and Python floats raise `AttributeError`.

## Impact
The MSMPR steady-state helper cannot run with scalar seeds, even though that path is explicitly implemented as a scalar root solve.

## Likely fix
Either pass a full concentration vector from `solve_steady_state`, or make `get_kinetics` handle scalar target concentrations when only one crystallizing component is being solved.

<!-- gh-triage-issue -->
## Triage: reproduced ✅

Root cause is exactly as diagnosed in the maintainer audit above: `MSMPR.solve_steady_state` ([Crystallizers.py:1775](https://github.com/SECQUOIA/PharmaPy/blob/master/PharmaPy/Crystallizers.py#L1775)) feeds the scalar `w_tank` from its `newton` root solve into `CrystKinetics.get_kinetics`, which unconditionally evaluates `conc.T[self.target_idx]` ([Kinetics.py:829](https://github.com/SECQUOIA/PharmaPy/blob/master/PharmaPy/Kinetics.py#L829)).

### Environment
- Base: `origin/master` @ `1c3a046` (cited `Kinetics.py`/`Crystallizers.py` byte-identical on the working branch).
- `python -m pytest` (numpy 2.2.6, scipy 1.15.3).

### Failing regression test (red on base)
```python
import numpy as np
from PharmaPy.Kinetics import CrystKinetics


def test_get_kinetics_accepts_scalar_target_concentration():
    kin = CrystKinetics(
        coeff_solub=[0.1, 0.0, 0.0],
        nucl_prim=[1.0, 0.0, 1.0],
        growth=[1.0, 0.0, 1.0],
        sup_sat_type='absolute',
    )
    kin.target_idx = 0

    w_tank = 0.5  # scalar seed, as produced by scipy.optimize.newton
    nucl, growth, _ = kin.get_kinetics(w_tank, 298.15, 0.5)

    assert np.ndim(nucl) == 0 and np.ndim(growth) == 0
    assert growth > 0
```
Red output:
```
    conc_target = conc.T[self.target_idx]
E   AttributeError: 'float' object has no attribute 'T'
```

### Observed vs expected
| input `conc` | observed | expected |
| --- | --- | --- |
| Python float (`newton` seed) | `AttributeError: 'float' object has no attribute 'T'` | scalar `nucl`, `growth` returned |
| numpy 0-d scalar | `IndexError` (too many indices) | scalar `nucl`, `growth` returned |

### Acceptance criteria
- `CrystKinetics.get_kinetics` returns finite scalar kinetics when given a scalar target concentration (the single-crystallizing-component case).
- `MSMPR.solve_steady_state(frac_seed, temp)` runs to a converged `w_convg` for a scalar `frac_seed` instead of raising.
- Existing vector/array `get_kinetics` callers (transient MSMPR/`SemibatchCryst` solves) are unaffected.

Next: `gh-issue-to-pr 61` — reuse this test and satisfy the criteria above.

--- a/PharmaPy/Kinetics.py
+++ b/PharmaPy/Kinetics.py
@@ -825,8 +825,17 @@ class CrystKinetics:
 
     def get_kinetics(self, conc, temp, kv_cry,
                      moments=None, nucl_sec_out=False):
+        """Evaluate crystallization kinetics for target concentration states.
 
-        conc_target = conc.T[self.target_idx]
+        Scalar concentrations are treated as the already-selected target
+        concentration, matching single-component steady-state solves.
+        """
+
+        conc_array = np.asarray(conc)
+        if conc_array.ndim == 0:
+            conc_target = conc_array.item()
+        else:
+            conc_target = conc_array.T[self.target_idx]
 
         # Supersaturation
         conc_sat = self.get_solubility(temp, conc)
@@ -838,6 +847,11 @@ class CrystKinetics:
         if self.sup_sat_type == 'ratio':
             sup_sat = sup_sat / conc_sat + 1
 
+        def is_default_secondary(name):
+            return (name == 'nucl_sec' and
+                    name not in self.custom_mechanisms and
+                    self.params_sec is None)
+
         par_p, par_s, par_g, par_d = self.params.values()
 
         if 'nucl_sec' in self.custom_mechanisms:
@@ -847,7 +861,12 @@ class CrystKinetics:
             args_sec = (kv_cry, self.reformulate_kin)
             par_sec = par_s
 
-        if isinstance(sup_sat, float):
+        if np.ndim(sup_sat) == 0:
+            sup_sat = np.asarray(sup_sat).item()
+            conc_sat = np.asarray(conc_sat).item()
+            if np.ndim(temp) == 0:
+                temp = np.asarray(temp).item()
+
             # print(sup_sat)
             args = [sup_sat, conc_sat, moments, temp, self.temp_ref]
             if sup_sat >= 0:
@@ -860,7 +879,9 @@ class CrystKinetics:
             mechs = {}
 
             for name in subset_mech:
-                if name in self.custom_mechanisms:
+                if is_default_secondary(name):
+                    mechs[name] = 0
+                elif name in self.custom_mechanisms:
                     args_concat = args + [self.params[name]]
                     mechs[name] = self.custom_mechanisms[name](*args_concat)
                 else:
@@ -907,7 +928,9 @@ class CrystKinetics:
                     temp_positive, self.temp_ref]
 
             for name in subset_mech:
-                if name in self.custom_mechanisms:
+                if is_default_secondary(name):
+                    continue
+                elif name in self.custom_mechanisms:
                     args_concat = args + [self.params[name]]
                     mechs[name][positive_map] = self.custom_mechanisms[name](*args_concat)
                 else:

--- a/PharmaPy/Kinetics.py
+++ b/PharmaPy/Kinetics.py
@@ -848,9 +848,15 @@ class CrystKinetics:
             sup_sat = sup_sat / conc_sat + 1
 
         def is_default_secondary(name):
-            return (name == 'nucl_sec' and
-                    name not in self.custom_mechanisms and
-                    self.params_sec is None)
+            """Return whether secondary nucleation is the inactive default."""
+            if name != 'nucl_sec' or name in self.custom_mechanisms:
+                return False
+
+            default = np.zeros_like(self.params['nucl_sec'], dtype=float)
+            if self.reformulate_kin:
+                default[0] = np.log(eps)
+
+            return np.allclose(self.params['nucl_sec'], default)
 
         par_p, par_s, par_g, par_d = self.params.values()
 

--- a/tests/test_cryst_kinetics.py
+++ b/tests/test_cryst_kinetics.py
@@ -81,6 +81,31 @@ def test_get_kinetics_preserves_vector_target_selection():
     np.testing.assert_allclose(dissol, [0.0, 0.0])
 
 
+def test_get_kinetics_uses_secondary_parameters_from_vector_update():
+    kin = _primary_growth_kinetics()
+    kin.set_params(
+        np.array([
+            1.0, 0.0, 1.0,
+            1.0, 0.0, 1.0, 1.0,
+            1.0, 0.0, 1.0,
+            0.0, 0.0, 0.0,
+        ])
+    )
+
+    conc = np.array([[0.5], [0.6]])
+    temp = np.array([298.15, 298.15])
+    moments = np.full((2, 4), 2.0)
+
+    prim, sec, growth, dissol = kin.get_kinetics(
+        conc, temp, 0.5, moments, nucl_sec_out=True
+    )
+
+    np.testing.assert_allclose(prim, [0.4, 0.5])
+    np.testing.assert_allclose(sec, [0.4, 0.5])
+    np.testing.assert_allclose(growth, [0.4, 0.5])
+    np.testing.assert_allclose(dissol, [0.0, 0.0])
+
+
 def test_msmpr_steady_state_accepts_scalar_seed(monkeypatch):
     MSMPR = _import_msmpr(monkeypatch)
 

--- a/tests/test_cryst_kinetics.py
+++ b/tests/test_cryst_kinetics.py
@@ -1,0 +1,117 @@
+import sys
+from types import ModuleType
+
+import numpy as np
+import pytest
+
+from PharmaPy.Kinetics import CrystKinetics
+
+
+pytestmark = pytest.mark.unit
+
+
+def _primary_growth_kinetics():
+    kin = CrystKinetics(
+        coeff_solub=[0.1, 0.0, 0.0],
+        nucl_prim=[1.0, 0.0, 1.0],
+        growth=[1.0, 0.0, 1.0],
+        sup_sat_type="absolute",
+    )
+    kin.target_idx = 0
+    return kin
+
+
+def _stub_assimulo_modules(monkeypatch):
+    assimulo = ModuleType("assimulo")
+
+    solvers = ModuleType("assimulo.solvers")
+    solvers.CVode = object
+
+    problem = ModuleType("assimulo.problem")
+    problem.Explicit_Problem = object
+
+    exception = ModuleType("assimulo.exception")
+    exception.TerminateSimulation = Exception
+
+    monkeypatch.setitem(sys.modules, "assimulo", assimulo)
+    monkeypatch.setitem(sys.modules, "assimulo.solvers", solvers)
+    monkeypatch.setitem(sys.modules, "assimulo.problem", problem)
+    monkeypatch.setitem(sys.modules, "assimulo.exception", exception)
+
+
+def _import_msmpr(monkeypatch):
+    try:
+        from PharmaPy.Crystallizers import MSMPR
+    except ModuleNotFoundError as exc:
+        if exc.name != "assimulo":
+            raise
+        _stub_assimulo_modules(monkeypatch)
+
+        from PharmaPy.Crystallizers import MSMPR
+
+    return MSMPR
+
+
+@pytest.mark.parametrize("conc", [0.5, np.array(0.5)])
+def test_get_kinetics_accepts_scalar_target_concentration(conc):
+    kin = _primary_growth_kinetics()
+
+    nucl, growth, dissol = kin.get_kinetics(conc, 298.15, 0.5)
+
+    assert np.ndim(nucl) == 0
+    assert np.ndim(growth) == 0
+    assert np.ndim(dissol) == 0
+    assert nucl == pytest.approx(0.4)
+    assert growth == pytest.approx(0.4)
+    assert dissol == pytest.approx(0.0)
+
+
+def test_get_kinetics_preserves_vector_target_selection():
+    kin = _primary_growth_kinetics()
+    kin.target_idx = 1
+
+    conc = np.array([[9.0, 0.5], [8.0, 0.6]])
+    temp = np.array([298.15, 298.15])
+    moments = np.zeros((2, 4))
+
+    nucl, growth, dissol = kin.get_kinetics(conc, temp, 0.5, moments)
+
+    np.testing.assert_allclose(nucl, [0.4, 0.5])
+    np.testing.assert_allclose(growth, [0.4, 0.5])
+    np.testing.assert_allclose(dissol, [0.0, 0.0])
+
+
+def test_msmpr_steady_state_accepts_scalar_seed(monkeypatch):
+    MSMPR = _import_msmpr(monkeypatch)
+
+    crystallizer = MSMPR.__new__(MSMPR)
+    crystallizer.vol_slurry = 1.0
+    crystallizer.target_ind = 0
+    crystallizer._Kinetics = _primary_growth_kinetics()
+
+    class Solid:
+        x_distrib = np.array([0.0, 1.0])
+        kv = 0.0
+
+        def getDensity(self, temp):
+            return 1.0
+
+    class Liquid:
+        mass_frac = np.array([0.5])
+
+    class Inlet:
+        vol_flow = 1.0
+        Liquid_1 = Liquid()
+
+    crystallizer.Solid_1 = Solid()
+    crystallizer._Inlet = Inlet()
+
+    x_vec, f_convg, w_convg, info, final_fn = crystallizer.solve_steady_state(
+        0.3, 298.15
+    )
+
+    np.testing.assert_allclose(x_vec, [0.0, 1.0])
+    np.testing.assert_allclose(f_convg, [1.0, np.exp(-2.5)])
+    assert w_convg == pytest.approx(0.5)
+    assert info.converged
+    assert final_fn == pytest.approx(0.0)

